src/main/java/graphql/execution/AsyncExecutionStrategy.java

```diff
@@ -39,22 +39,29 @@ public AsyncExecutionStrategy(DataFetcherExceptionHandler exceptionHandler) {
         Map<String, List<Field>> fields = parameters.fields();
         List<String> fieldNames = new ArrayList<>(fields.keySet());
         List<CompletableFuture<ExecutionResult>> futures = new ArrayList<>();
-        for (String fieldName : fieldNames) {
-            List<Field> currentField = fields.get(fieldName);
+        try {
+            for (String fieldName : fieldNames) {
+                List<Field> currentField = fields.get(fieldName);
 
-            ExecutionPath fieldPath = parameters.path().segment(fieldName);
-            ExecutionStrategyParameters newParameters = parameters
-                    .transform(builder -> builder.field(currentField).path(fieldPath));
+                ExecutionPath fieldPath = parameters.path().segment(fieldName);
+                ExecutionStrategyParameters newParameters = parameters
+                        .transform(builder -> builder.field(currentField).path(fieldPath));
 
-            CompletableFuture<ExecutionResult> future = resolveField(executionContext, newParameters);
-            futures.add(future);
+                CompletableFuture<ExecutionResult> future = resolveField(executionContext, newParameters);
+                futures.add(future);
+            }
+        } catch (AbortExecutionException abortExecution) {
```

Comment:
Not sure when this exception can happen at all: I would probably to except that we need to check the futures.

Reply:
Something like `CompletableFuture.anyOf(futures)` ...

Reply:
The use of AbortExecutionException is our special exception type.

That is if your Instrumentation / DataFetcher throws this, then it instructs us to do a partial result (if you have entered execution proper that is)

I did another commit that adds the DF support

So it becomes a special "signal" to stop execution once it has begun.

`CompletableFuture.anyOf(futures)` is no good since it waits for any one of the outstanding work to be done.  Whereas we want immediate return I think

Reply:
Sorry, I was not clear. 

It was more about the style: I don't think a method returning a `CF` should also throw exceptions: It should return a `CF` which completed exceptionally.  (Which means we should catch sync exceptions inside of `fetchField` etc)

Because otherwise we are mixing async with sync code, which we should try to avoid. 

So I don't think we should catch a exception here, but handle failed `CF` results. (But you are right: not with `.anOf` ... we have to deal with each `CF` separately).

For example one case I think is still missing: a DataFetcher returning a `CF` which completed exceptionally with a `AbortExecutionException`.

Reply:
Another thought: maybe even `Instrumentation` itself should become async too ... so that you can return a `CF` from `beginField` for example.

Reply:
Ok I get your point better.   I will have another go at it using CFs and understanding the exceptions inside the CFs.

As for the case of a  DataFetcher returning a CF which completed exceptionally with a AbortExecutionException this would not be possible (in AsyncExecutionStrategy) since its not shortcutting anything.  We would not known until it completes that we should shortcut execution and hence we would wait for .join() and it would be all to late

Not sure CFs with Intrumentation makes that much sense.  For things like "timing and logging" you want it to happen at the time we start and actively complete once the inner CF completes.  A CF inside a CF is too hard to co-ordinate I think

Reply:
Ok looking into this more, I can see how we can do it for AsyncSerialES (because it does them one at a time and hence can short circuit) but I cant see how you can do it for AsyncES

Its code is basically

            for (String fieldName : fieldNames) {
                List<Field> currentField = fields.get(fieldName);
                CompletableFuture<ExecutionResult> future = resolveField(executionContext, newParameters);
                futures.add(future);
            }

and then 

    Async.each(futures).whenComplete(handleResults(executionContext, fieldNames, overallResult));

So you have fired off N fields at one time and then you wait for them all the complete.

But partial result with shortcircuiting is the opposite of that.  You need the call to  resolveField(executionContext, newParameters) to be able to tell you that it wants to shortcircuit.

Exceptions are that short circuit.  You could invent some dual type that is `Either<CF<ExeutionResult>,ShortCircuitInfo>` but really that's what exception is in Java

I dont think you can "interupt" execution (while capturing a paertial result) without using exceptions